### PR TITLE
Initialized M on cell faces

### PR DESCRIPTION
**Pull request by prkkumar**

### CODE REVIEW

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp [hunk 1]
```diff
@@ -69,7 +72,7 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
     // temporary Multifab storing M from previous timestep (old_time) before updating to M(new_time)
     std::array<std::unique_ptr<amrex::MultiFab>, 3> Mfield_old; // Mfield_old is M(old_time)
 
-    amrex::GpuArray<int, 3> const& mag_Ms_stag         = macroscopic_properties->mag_Ms_IndexType;
+    //amrex::GpuArray<int, 3> const& mag_Ms_stag         = macroscopic_properties->mag_Ms_IndexType;
```
Comment (prkkumar) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp]:
```suggestion
```

**Comment by prkkumar** [comment]
This PR is to 

1. Initialize M directly on the cell faces using user-defined external grid function
2. In the `MacroscopicEvolveHM` routines, instead of using cell-centered Ms averaged to face-value, Ms is computed directly where M is updated.

This will refactor M and Ms and ensure that it is ready for use in H_exchange boundary conditions where we need to define Dirichlet and Neumann BCs.

As a verification of the implementation, "attenuation spectrum of the magnetically tunable filter" test case using this approach is compared with the previous approach and the results are in good agreement as shown below:

**Comment by prkkumar** [comment]
image

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp [hunk 2]
```diff
@@ -138,11 +142,18 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
         amrex::Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
         amrex::Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
 
+        int lev = 0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp]: Instead of hard-coding lev=0 here, Can you send the lev information from where this is called
`m_fdtd_solver_fp[lev]->MacroscopicEvolveHM(...,lev)`

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp [hunk 3]
```diff
@@ -81,6 +81,9 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian_2nd(
     amrex::GpuArray<int, 3> const& Mx_stag        = macroscopic_properties->Mx_IndexType;
     amrex::GpuArray<int, 3> const& My_stag        = macroscopic_properties->My_IndexType;
     amrex::GpuArray<int, 3> const& Mz_stag        = macroscopic_properties->Mz_IndexType;
+    amrex::GpuArray<int, 3> const& Bx_stag        = macroscopic_properties->Bx_IndexType;
+    amrex::GpuArray<int, 3> const& By_stag        = macroscopic_properties->By_IndexType;
+    amrex::GpuArray<int, 3> const& Bz_stag        = macroscopic_properties->Bz_IndexType;
```
Comment (jackieyao0114) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp:88]: why is `B_stag` needed given that `M_stag` has already been read in?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp [hunk 4]
```diff
@@ -78,6 +78,12 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
     amrex::GpuArray<int, 3> const& My_stag             = macroscopic_properties->My_IndexType;
     amrex::GpuArray<int, 3> const& Mz_stag             = macroscopic_properties->Mz_IndexType;
     amrex::GpuArray<int, 3> const& macro_cr            = macroscopic_properties->macro_cr_ratio;
+    amrex::GpuArray<int, 3> const& Hx_stag             = macroscopic_properties->Hx_IndexType;
+    amrex::GpuArray<int, 3> const& Hy_stag             = macroscopic_properties->Hy_IndexType;
+    amrex::GpuArray<int, 3> const& Hz_stag             = macroscopic_properties->Hz_IndexType;
+    amrex::GpuArray<int, 3> const& Bx_stag             = macroscopic_properties->Bx_IndexType;
+    amrex::GpuArray<int, 3> const& By_stag             = macroscopic_properties->By_IndexType;
+    amrex::GpuArray<int, 3> const& Bz_stag             = macroscopic_properties->Bz_IndexType;
```
Comment (jackieyao0114) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp:88]: `H_stag` is the same to `B_stag`, isn't it?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp [hunk 5]
```diff
@@ -138,11 +142,18 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
         amrex::Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
         amrex::Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
 
+        int lev = 0;
+        const auto dx = warpx.Geom(lev).CellSizeArray();
```
Comment (jackieyao0114) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp:148]: this is just my own semantic preference. Since `dx` contains three components `dx`, `dy` and `dz`, naming the array as `dx` is a bit odd to me. Maybe `d_cell`?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp [hunk 6]
```diff
@@ -965,8 +1016,17 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian_2nd(
 
                         [=] AMREX_GPU_DEVICE(int i, int j, int k) {
 
-                            Real mag_Ms_arrz = MacroscopicProperties::macro_avg_to_face(i,j,k,Mzface_stag,mag_Ms_arr);
-
+                            amrex::Real x, y, z;
+                            x = problo[0] + i*dx[0] + (1._rt - Mzface_stag[0]) * dx[0]*0.5;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp]: can you call GetCellCoordinates() here too. so we reduce code duplication

**Review by RevathiJambunathan — COMMENTED** [review]
Thank you for this PR @prkkumar ! yay first PR!
just two minor comments for now. 
will do a detailed review later today

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp (hunk 5 above)
Comment (RevathiJambunathan) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp:148]: :) yes -- I agree that "dx" has "dx,dy,dz" for 3D and that its weird :)

But `dx=CellSizeArray()` is consistent with AMReX usage (https://amrex-codes.github.io/amrex/docs_html/Basics.html#realbox-and-geometry), so its used that way in warpx too.
So for consistency, we can choose to have the same usage.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp (hunk 6 above)
Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp]: Yes, addressed this.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp (hunk 2 above)
Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp]: yes, addressed this.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp (hunk 5 above)
Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp:148]: Going with the AMReX convention, I am sticking with dx.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp (hunk 3 above)
Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp:88]: To keep the code generic, we go with separate staggerings for B, H, and M updates.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp (hunk 4 above)
Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp:88]: Yes they are the same in our case, but to keep the code generic we can go with separate stageerings for B and H updates.

**Review by ajnonaka — APPROVED** [review]

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp [hunk 7]
```diff
@@ -138,11 +145,17 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian(
         amrex::Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
         amrex::Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
 
+        const auto dx = warpx.Geom(lev).CellSizeArray();
+        const auto problo = warpx.Geom(lev).ProbLoArray();
+        const auto macro_parser = macroscopic_properties->m_mag_Ms_parser->compile<3>();
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM.cpp]: may be call this 'mag_parser' since macro could mean many things?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H [hunk 8]
```diff
@@ -54,6 +54,18 @@ public:
...
+     /** Gpu Vector with index type of the Hz multifab */
+     amrex::GpuArray<int, 3> Hz_IndexType;
+     /** Gpu Vector with index type of the Bx multifab */
+     amrex::GpuArray<int, 3> Bx_IndexType;
+     /** Gpu Vector with index type of the By multifab */
+     amrex::GpuArray<int, 3> By_IndexType;
+     /** Gpu Vector with index type of the Bz multifab */
+     amrex::GpuArray<int, 3> Bz_IndexType;
      /** Gpu Vector with index type of the saturation magnetization multifab */
      amrex::GpuArray<int, 3> mag_Ms_IndexType;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H]: If mag_Ms_IndexType is not used anymore, we can delete this too.
Also remove the function `getmax_Ms_mf()`
as well as the mag_Ms multifab.

Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H]: Done. Thanks!

File: Source/Utils/WarpXUtil.H [hunk 9]
```diff
@@ -194,6 +194,38 @@ void getCellCoordinates (int i, int j, int k,
...
+ * \param[in] k    index along z
+ * \param[in] mf_type IntVect containing the staggering type to convert (i,j,k) to (x,y,z)
+ * \param[in] domain_lo Physical coordinates of the lowest corner of the simulation domain
+ * \param[in] dx   Cell size of the simulation domain
+ *
+ * \param[out] x   physical coordinate along x
+ * \param[out] y   physical coordinate along y
+ * \param[out] z   physical coordinate along z
+ */
+AMREX_GPU_DEVICE AMREX_FORCE_INLINE
```
Comment (RevathiJambunathan) [inline Source/Utils/WarpXUtil.H:211]:
```suggestion
AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```

**Review by RevathiJambunathan — APPROVED** [review]
Thank you for the PR, @prkkumar !

File: Source/Utils/WarpXUtil.H [hunk 10]
```diff
@@ -194,6 +194,38 @@ void getCellCoordinates (int i, int j, int k,
...
+void getCellCoordinatesIntVect (int i, int j, int k,
```
Comment (RevathiJambunathan) [inline Source/Utils/WarpXUtil.H]: what is the difference between this and the above get CellCoordinates?

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp [hunk 11]
```diff
@@ -943,8 +977,9 @@ void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian_2nd(
 
                         [=] AMREX_GPU_DEVICE(int i, int j, int k) {
 
-                            Real mag_Ms_arry = MacroscopicProperties::macro_avg_to_face(i,j,k,Myface_stag,mag_Ms_arr);
-
+                            amrex::Real x, y, z;
+                            WarpXUtilAlgo::getCellCoordinatesIntVect(i, j, k, Myface_stag, problo, dx, x, y, z);
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp]: We could use the GpuArray for My_stag here as well?

File: Source/Utils/WarpXUtil.H (hunk 10 above)
Comment (prkkumar) [reply Source/Utils/WarpXUtil.H]: Discussed this during meeting. I have removed this function, and I am using getCellCoordinates function above throughout this PR.

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp (hunk 11 above)
Comment (prkkumar) [reply Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp]: Discussed this during the meeting. Since Myface_stag is same as My_stag, we can use the same getCellCoordinate function for GPU array.

**Review by RevathiJambunathan — APPROVED** [review]
